mm2src/coins/eth/eth_swap_v2/mod.rs

```diff
@@ -108,6 +120,112 @@ impl EthCoin {
             EthCoinType::Nft { .. } => Err("NFT protocol is not supported for ETH and ERC20 Swaps".to_string()),
         }
     }
+
+    /// A helper function that scans blocks for a specific event containing the given `swap_id`,
+    /// returning transaction hash of spend transaction once found.
+    /// **NOTE:** The current function implementation assumes that `swap_id` is the first 32 bytes of the transaction input data.
+    pub(crate) async fn find_transaction_hash_by_event(
+        &self,
+        params: SpendTxSearchParams<'_>,
+    ) -> MmResult<H256, FindPaymentSpendError> {
+        loop {
+            let now = now_sec();
+            if now > params.wait_until {
+                return MmError::err(FindPaymentSpendError::Timeout {
+                    wait_until: params.wait_until,
+                    now,
+                });
+            }
+
+            let current_block = match self.current_block().compat().await {
+                Ok(b) => b,
+                Err(e) => {
+                    error!("Error getting block number: {}", e);
+                    Timer::sleep(params.check_every).await;
+                    continue;
+                },
+            };
+
+            let mut next_from_block = params.from_block;
+            while next_from_block <= current_block {
+                let to_block = std::cmp::min(next_from_block + self.logs_block_range - 1, current_block);
+
+                // Fetch events for the current block range
+                let events = match self
+                    .events_from_block(
+                        params.swap_contract_address,
+                        params.event_name,
+                        next_from_block,
+                        Some(to_block),
+                        params.abi_contract,
+                    )
+                    .await
+                {
+                    Ok(events) => events,
+                    Err(e) => {
+                        error!(
+                            "Error getting {} events from {} to {} block: {}",
+                            params.event_name, next_from_block, to_block, e
+                        );
+                        Timer::sleep(params.check_every).await;
+                        // Move to next window if there was an error
+                        next_from_block += self.logs_block_range;
+                        continue;
```

Comment:
why? we might skip important blocks this way.

Reply:
hm I thought that if an error occurred, we could try other blocks. If they don't have an event, we might start the process again from the beginning of the outer `loop` and repeat the blocks. but time could run out, so it might never happen, which is bad. will remove it

Reply:
done https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261/commits/a9f481ce5dffaf3c1e9b86221812e12e5f238db1